ngraph/core/reference/include/ngraph/runtime/reference/proposal.hpp

```diff
@@ -0,0 +1,445 @@
+                    {
+                        for (int i = 5 * num_rois; i < 5 * post_nms_topn_; i++)
+                        {
+                            rois[i] = static_cast<T>(0.f);
+                        }
+                    }
+                }
+            } // namespace details
+
+            template <typename T>
+            void proposal(const T* class_probs,
+                          const T* bbox_deltas,
+                          const T* image_shape,
+                          T* output,
+                          const Shape& class_probs_shape,
+                          const Shape& bbox_deltas_shape,
+                          const Shape& image_shape_shape,
+                          const Shape& output_shape,
+                          const op::ProposalAttrs& attrs)
+            {
+                const T* p_bottom_item = class_probs;
+                const T* p_d_anchor_item = bbox_deltas;
+                T* p_roi_item = output;
+                // bottom shape (batch_size * (2 * num_anchors) * H * W)
+                const unsigned int bottom_H = class_probs_shape[2];
+                const unsigned int bottom_W = class_probs_shape[3];
+                // input image height and width
+                const T img_H = image_shape[0];
+                const T img_W = image_shape[1];
+                // scale factor for H and W, depends on shape of image_shape
+                // can be split into H and W {image_height, image_width, scale_height, scale_width}
+                // or be the same for both   {image_height, image_width, scale_height_and_width}
+                const T scale_H = image_shape[2];
+                const T scale_W = (image_shape_shape.size() < 4 ? scale_H : image_shape[3]);
+                const T min_box_H = attrs.min_size * scale_H;
+                const T min_box_W = attrs.min_size * scale_W;
+                // get number of proposals
+                // class_probs shape is {batch_size, anchor_count*2, bottom_H, bottom_W}
+                const unsigned int anchor_count = class_probs_shape[1] / 2;
+                const unsigned int num_proposals = anchor_count * bottom_H * bottom_W;
+                // final RoI count
+                int num_rois = 0;
+                std::vector<details::ProposalBox<T>> proposals(num_proposals);
+                const int pre_nms_topn =
+                    num_proposals < attrs.pre_nms_topn ? num_proposals : attrs.pre_nms_topn;
+                std::vector<int> is_dead(pre_nms_topn);
+                std::vector<unsigned int> roi_indices(attrs.post_nms_topn);
+
+                std::vector<float> anchors(4 * anchor_count);
+                details::generate_anchors(attrs, anchors);
+
+                unsigned int batch_num = class_probs_shape[0];
+                float coordinates_offset = attrs.framework == "tensorflow" ? 0.0f : 1.0f;
+                bool initial_clip = attrs.framework == "tensorflow" ? true : false;
```

Comment:
Expression `attrs.framework == "tensorflow"` return boolean type.

Reply:
Yes it does, but I wanted to indicate here that those framework dependent variable values don't have anything to do with the comparison result, only with the framework used (maybe it does not make much sense the way I put it). The other point is now those assignments are uniform for all variable types. So I'll leave it as is is.